src/v/v8_engine/script.cc

```diff
@@ -30,16 +39,46 @@ script::script(size_t max_heap_size_in_bytes) {
       0, max_heap_size_in_bytes);
     _isolate = std::unique_ptr<v8::Isolate, isolate_deleter>(
       v8::Isolate::New(isolate_params), isolate_deleter());
+
+    _watchdog.set_callback([this] { stop_execution(); });
 }
 
 script::~script() {
     _function.Reset();
     _context.Reset();
 }
 
-void script::init(std::string_view name, ss::temporary_buffer<char> js_code) {
-    compile_script(std::move(js_code));
-    set_function(name);
+ss::future<>
+script::init(ss::sstring name, ss::temporary_buffer<char> js_code) {
```

Comment:
`init()` and `run()` are very similar you can break out similar parts into a third function to reduce code duplication

Reply:
The used different function(
init uses `Compile` and `Run` from v8
run uses `Call` from v8 (haha, it is so crazy)